src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,170 @@
+package org.rust.ide.annotator
+
+import com.google.gson.JsonParser
+import com.intellij.lang.annotation.*
+import com.intellij.lang.annotation.Annotation
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.module.ModuleUtilCore
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.ModificationTracker
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import com.intellij.util.PathUtil
+import org.apache.commons.lang.StringEscapeUtils.escapeHtml
+import org.rust.cargo.project.settings.toolchain
+import org.rust.cargo.toolchain.CargoMessage
+import org.rust.cargo.toolchain.CargoSpan
+import org.rust.cargo.toolchain.CargoTopMessage
+
+data class CargoCheckAnnotationInfo(val file: PsiFile, val editor: Editor)
+
+class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Project? = null) : ModificationTracker {
+
+    var messages = emptyList<CargoTopMessage>()
+
+    init {
+        val parser = JsonParser()
+        messages += commandOutput.filter {
+            // Apparently some lines also have human readable messages. We ignore any lines not beginning with '{'
+            val trimmed = it.trimStart()
+            trimmed.isNotEmpty() && trimmed[0] == '{'
+        }.filter {
+            // We check the first field name by finding the first " then reading until another " (we expect
+            // that a '{' present and don't check for it).
+            val first = it.indexOfFirst { it == '"' }
+            assert(first != -1)
+            val second = it.indexOf('"', first + 1)
+            assert(second != -1)
+
+            when (it.subSequence(first + 1, second)) {
+                "message" -> true
+                "features" -> false
+                else -> {
+                    assert(false)
+                    throw AssertionError()
+                }
+            }
+        }.map {
+            parser.parse(it)
+        }.mapNotNull {
+            if (it.isJsonObject)
+                CargoTopMessage.fromJson(it.asJsonObject)
+            else
+                null
+        }
+    }
+
+    override fun getModificationCount() = project?.baseDir?.modificationStamp ?: 0
```

Comment:
I think it's a wrong way to determine that anything in the project has been changed. What's the right way?